nexus/src/db/datastore.rs

```diff
@@ -1944,6 +1946,148 @@ impl DataStore {
             .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
+    /// Update a network interface associated with a given instance.
+    pub async fn instance_update_network_interface(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        authz_interface: &authz::NetworkInterface,
+        updates: NetworkInterfaceUpdate,
+    ) -> UpdateResult<NetworkInterface> {
+        use crate::db::schema::network_interface::dsl;
+
+        // This database operation is surprisingly subtle, and I don't believe
+        // it's possible to express in a single query.
+        //
+        // First, we need to check whether the instance is stopped. We do that
+        // in other cases with a CTE designed to fail if that's not true, but
+        // we're forced into a transaction by other considerations below.
+        //
+        // More importantly, setting a new primary interface is a pretty complex
+        // operation, and depends on the identity of the target. There are four
+        // cases, the truth table for (make_primary, target_is_primary)
+        //  - (false, _) => In this case, we only need to update a single row,
+        //  and we can apply the updates requested, minus the make_primary
+        //  value, unconditionally. This is pretty straightforward.
+        //
+        //  - (true, true) => This is also pretty simple, since we can just
+        //  apply the updates directly.
+        //
+        //  - (true, false) => This one is complicated. We need to apply all
+        //  updates to the target row, and then set the `is_primary` column of
+        //  the old primary to false. This isn't a single query.
+        //
+        //  The issue is that for the latter two cases, we don't know which
+        //  we're taking until we make a query to the database. We need to
+        //  figure out if the target is the primary first, and then apply one or
+        //  two updates accordingly.
+        //
+        //  This is all pretty complicated, with a number of branches and
+        //  additional consistency checks, such as the fact that the instance
+        //  has to be stopped. For now, we'll just bite the bullet and run this
+        //  all in a transaction.
```

Comment:
I think this is the right approach, fwiw. We have enough stuff to build, there will be space to optimize if we need that (and that's easier when cases like this are well-documented, like this is!)